application/common/models/User.php

```diff
@@ -1003,8 +1025,46 @@ public function isPromptForMfa(): bool
         return false;
     }
 
-    public function updateExternalGroups($appPrefix, $appExternalGroups): bool
+    public static function syncExternalGroups(string $appPrefix, array $desiredExternalGroupsByUserEmail): array
     {
+        $errors = [];
+        $emailAddressesOfCurrentMatches = self::listUsersWithExternalGroupWith($appPrefix);
+
+        // Indicate that users not in the "desired" list should not have any
+        // such external groups.
+        foreach ($emailAddressesOfCurrentMatches as $email) {
+            if (! array_key_exists($email, $desiredExternalGroupsByUserEmail)) {
+                $desiredExternalGroupsByUserEmail[$email] = '';
+            }
+        }
+
+        foreach ($desiredExternalGroupsByUserEmail as $email => $groupsForPrefix) {
+            $user = User::findByEmail($email);
+            if ($user === null) {
+                $errors[] = 'No user found for email address ' . json_encode($email);
+                continue;
+            }
+            $successful = $user->updateExternalGroups($appPrefix, $groupsForPrefix);
```

Comment:
I could be wrong, but the `app` part of the `appPrefix` parameter name seems out of place. It may have meaning elsewhere, but it just raises questions in my mind. My 2 cents.

Reply:
Yeah, I've been having trouble deciding on a good name for that. It's the group-name prefix for a particular application (e.g. "wiki").

This `groups_external` field can hold groups used by multiple different applications (SPs). For example, a User's `groups_external` field might have `wiki-managers,map-america,map-europe` (which would be a "wiki-managers" group for use by the "wiki" application, and "map-america" and "map-europe" to be used by the "map" application, or at least that's the intent behind this design.

If given enough context, I can simply call it the prefix... and perhaps I should just use that everywhere.

Thanks for sharing your 2 cents. That's often helpful.